pkg/util/idlocker.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 ceph-csi authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"sync"
+)
+
+/*
+IDLock is a per identifier lock with a use counter that retains a number of users of the lock.
+IDLocker is a map of IDLocks holding the IDLocks based on a passed in identifier.
+Typical usage (post creating an IDLocker) is to Lock/Unlock based on identifiers as per the API.
+*/
+type (
+	IDLock struct {
+		mtx      sync.Mutex
+		useCount int
+	}
+
+	IDLocker struct {
+		lMutex sync.Mutex
+		lMap   map[string]*IDLock
+	}
+)
+
+func NewIDLocker() *IDLocker {
+	return &IDLocker{
+		lMap: make(map[string]*IDLock),
+	}
+}
+
+func (lkr *IDLocker) Lock(identifier string) *IDLock {
+	var lk *IDLock
+	newlk := new(IDLock)
+
+	lkr.lMutex.Lock()
+	lk = lkr.lMap[identifier]
+	if lk == nil {
+		lk = newlk
+		lkr.lMap[identifier] = lk
+	}
+	lk.useCount++
```

Comment:
am failing to understand what's the use of `useCount`?

Reply:
now got it. with the use of `useCount` which helps in avoiding the key deleting from the map (any performance improvement)?

Reply:
Right, UseCount is to delete when there are no more users.

Performance improvement as in due to UseCount or key deletion? IMO none.

Reply:
we have to do performance measurement, it can be done in other PR
i want to use sync.Map and check which one performs better